tests/phpunit/CRM/Contribute/BAO/ContributionTest.php

```diff
@@ -1688,4 +1688,91 @@ public function testCancelOrderWithPledge(): void {
     $this->assertCount(0, $payments);
   }
 
+  /**
+   * Test activity contact is updated when contribution contact is changed
+   */
+  public function testUpdateActivityContactOnContributionContactChange(): void {
+    $contactId_1 = $this->individualCreate();
+    $contactId_2 = $this->individualCreate();
+    $contactId_3 = $this->individualCreate();
+
+    $contributionParams = [
+      'financial_type_id' => 'Donation',
+      'receive_date' => date('Y-m-d H:i:s'),
+      'sequential' => TRUE,
+      'total_amount' => 50,
+    ];
+
+    // Case 1: Only source contact, no target contact
+
+    $contribution = $this->callAPISuccess('Contribution', 'create', array_merge(
+      $contributionParams,
+      ['contact_id' => $contactId_1]
+    ))['values'][0];
+
+    $activity = $this->callAPISuccess('Activity', 'get', [
```

Comment:
You don't need to change this but one small shortcut is 

```
$activity = $this->callAPISuccessGetSingle('Activity', ['source_record_id' => $contribution['id']])
```

of course you have to set a limit if there could be legimately more than one...